app/src/main/java/Domoserv_Pi/domoserv_android/MainActivity.kt

```diff
@@ -37,43 +38,129 @@ class MainActivity : AppCompatActivity() {
+        dialog.setCancelable(true)
+        dialog.setContentView(R.layout.set_state_mode_dialog)
+        val body = dialog.findViewById(R.id.title) as TextView
+        body.text = title
+
+        val zone = when(title.contains("1")) {
+            true -> Zone.Z1.ordinal
+            false -> Zone.Z2.ordinal
+        }
+
+        val spinner = dialog.findViewById(R.id.select) as Spinner
+        val dataAdapter = when(type) {
+            1 -> ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, modeList)
+            else -> ArrayAdapter<String>(this, android.R.layout.simple_spinner_item, stateList)
+        }
+        spinner.adapter = dataAdapter
+        spinner.setSelection(selected)
+
+        val submit = dialog.findViewById(R.id.submit) as Button
+        submit.setOnClickListener {
+            val index = dialog.findViewById<Spinner>(R.id.select).selectedItemPosition
+            when(type) {
+                1 -> changeMode(zone, index)
+                else -> changeState(zone, index)
+            }
+            dialog.dismiss()
+        }
+        dialog.show()
+    }
+
+    private fun changeState(zone: Int, state: Int) {
+        ws.send(getString(R.string.setZoneState).replace("{zone}",zone.toString()).replace("{state}",state.toString()))
+        startUpdate()
+    }
+
+    private fun changeMode(zone: Int, mode: Int) {
+        ws.send(getString(R.string.setZoneMode).replace("{zone}",zone.toString()).replace("{mode}",mode.toString()))
+        startUpdate()
     }
 
-    @RequiresApi(Build.VERSION_CODES.O)
     fun startUpdate() {
-        ws.send("CVOrder|GetZ1Order")
-        ws.send("CVOrder|GetZ2Order")
-        ws.send("CVOrder|GetZ1Status")
-        ws.send("CVOrder|GetZ2Status")
-        ws.send("CVOrder|GetRemainingTimeZ1")
-        ws.send("CVOrder|GetRemainingTimeZ2")
-        ws.send("CVOrder|GetTemp;0")
-        ws.send("CVOrder|GetTemp;1")
+        ws.send(getString(R.string.getZ1State))
+        ws.send(getString(R.string.getZ2State))
+        ws.send(getString(R.string.getZ1Mode))
+        ws.send(getString(R.string.getZ2Mode))
+        ws.send(getString(R.string.getZ1RemainingTime))
+        ws.send(getString(R.string.getZ2RemainingTime))
+        ws.send(getString(R.string.getIndoorTemp))
+        ws.send(getString(R.string.getOutdoorTemp))
         val date = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
+        val endDate = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
         val dateYear = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy")) + "-01-01"
-        ws.send("CVOrder|GetDataCPTEnergy;$dateYear:$date")
+        ws.send(getString(R.string.getDataEnergy).replace("{date}",date).replace("{endDate}",endDate))
     }
 
     fun updateField(data: String) {
-        if (data.contains("CVOrder")) {
-            if (data.contains("GetZ1Order")) {
+        if (data.contains(getString(R.string.cvorderValue))) {
```

Comment:
Je suis sûr qu'il y a moyen de factoriser tout ces if mais j'ai pas trouvé comment 👅
Peut-être voir avec des maps ... (mapOf<String, String> (key1 to value1, key2 to value2 ...))

Reply:
oui je trouve ça lourd aussi comme code mais je n'ai pas trouvé non plus, du moins pour le moment